### ISSUE TITLE
bundler 1.7.7 changed source dependency API support check behavior

### ISSUE THREAD

**Issue by dpiddy** [bundler/bundler#3296]
I have a `Gemfile` with a `source` hosted at S3. It does not support the dependency API and only makes available the files necessary to support index-based usage. All other requests return a 403.

Starting with bundler 1.7.7, my source no longer works. I get an error like this:

```
Authentication is required for https://my-s3-source.local
Please supply credentials for this source. You can do this by running:
 bundle config https://my-s3-source.local username:password
```

I believe it is caused by [this rescue change](https://github.com/bundler/bundler/compare/v1.7.6...v1.7.7#diff-23f4d7633f08ecf9c4a66f7189e34f0cR280).

**Comment by dpiddy** [comment]
Any update on this?

**Comment by TimMoore** [comment]
I haven't had a chance to look into this yet, sorry.

**Comment by hotgazpacho** [comment]
We're running into this as well, with sources pointing to internal geminabox servers.

**Comment by schneems** [comment]
This is a blocker to upgrading Heroku to 1.7.x any ideas what caused the regression?

**Comment by TimMoore** [comment]
@schneems the regression was caused by #3260, which fixed a problem using Gemfury as a private gem server. I can revert that change, but it will break things for other users. I would expect that to also block upgrading Heroku.

Fundamentally, the issue is that different gem servers have inconsistent behaviours with the meaning of a 403 status code. Some servers return it when the request is not authenticated, and re-sending the request with authentication fixes the problem, others return it to mean that they don't support the dependency API at all.

If anyone else would like to try a fix, that would be helpful... I haven't had much time for Bundler work recently.

**Comment by dpiddy** [comment]
A start might be to explicitly document bundler's expectations around sources, considering things like:

* when the dependency API will be used
* how dependency API support is checked
* when falling back to a full index will happen

That would let gem repo providers know what to do to be compatible.

Also how about the ability to force behavior per-source? For example, if I know a source should only use the full index, perhaps I could specify that as an option to `source`.

**Comment by TimMoore** [comment]
> Also how about the ability to force behavior per-source? For example, if I know a source should only use the full index, perhaps I could specify that as an option to source.

Yeah, I think that would be a good idea... even if everything were working correctly, it would save an extra request.

**Comment by TimMoore** [comment]
I have a proposed fix for this over at #3342. If anyone who has run into this problem can test that branch and report back, that would be very helpful. The more the better, because I've seen a variety of different behaviors from different servers.

**Comment by TimMoore** [comment]
Fixed in 1.7.12.
